runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -713,6 +705,86 @@ private void emitResults() {
     }
   }
 
+  private DoFnRunner<InputT, OutputT> ensureStateCleanup(
+      SdkHarnessDoFnRunner<InputT, OutputT> sdkHarnessRunner) {
+    if (keyCoder == null) {
+      // There won't be any state to clean up
+      // (stateful functions have to be keyed)
+      return sdkHarnessRunner;
+    }
+    // Takes care of state cleanup via StatefulDoFnRunner
+    Coder windowCoder = windowingStrategy.getWindowFn().windowCoder();
+    StatefulDoFnRunner.CleanupTimer<InputT> cleanupTimer =
+        new StatefulDoFnRunner.CleanupTimer<InputT>() {
+
+          private static final String GC_TIMER_ID = "__user-state-cleanup__";
+
+          @Override
+          public Instant currentInputWatermarkTime() {
+            return timerInternals.currentInputWatermarkTime();
+          }
+
+          @Override
+          public void setForWindow(InputT input, BoundedWindow window) {
+            Preconditions.checkNotNull(input, "Null input passed to CleanupTimer");
+            // make sure this fires after any window.maxTimestamp() timers
+            Instant gcTime = LateDataUtils.garbageCollectionTime(window, windowingStrategy).plus(1);
+            ByteBuffer key;
+            try {
+              key =
+                  ByteBuffer.wrap(
+                      CoderUtils.encodeToByteArray((Coder) keyCoder, ((KV) input).getKey()));
```

Comment:
Do we need to handle key collision in different windows?

Reply:
There is a timer for each window and key. The key has to be set here because Flink's state backend partitions state by key. Does that answer your question?